### PR TITLE
Add library syntax and testing

### LINKED ISSUE #478: Plugin libraries
1. All question name should be human readable
2. Only provide the validated version
3. The Matrix is already in the Main branch

Questionnaires:
- [x] vlat (components to test participants' visual literacy): https://ieeexplore.ieee.org/document/7539634
- [x] mini-vlat (components to test participants' visual literacy): https://onlinelibrary.wiley.com/doi/full/10.1111/cgf.14809
- [x] nasa-tlx (components to test for participant workload): https://humansystems.arc.nasa.gov/groups/tlx/
- [x] beauvis (a library for evaluating the aesthetics of visualizations): https://ieeexplore.ieee.org/document/9903341
- I cannot create a Matrix/Likert question
- [x] previs (Perceived Readability Evaluation for Visualizations scale): https://ieeexplore.ieee.org/document/10681245
- [x] SUS (System Usability Scale) https://digital.ahrq.gov/sites/default/files/docs/survey/systemusabilityscale%2528sus%2529_comp%255B1%255D.pdf
- [x] QUIS (Questionnaire for User Interaction Satisfaction): https://www.researchgate.net/publication/2624491_Assessing_the_Equivalence_of_the_Paper_and_On-line_Formats_of_the_QUIS_55
- [x] SMEQ (Subjective Mental Effort Questionnaire)
- Matrix with anchors on both side? https://revisit.dev/study/test-likert-matrix/
- Can I have a title / instruction text for the QuestionnaireComponent?
- Labels 0-9 rather than 1-10
- Points can't be aligned
- [x] SAM (Self Assessment Manikins): https://doi.org/10.1016/0005-7916(94)90063-9
- Need to use images as labels (an uncommon case)
- [x] berlin-num (Berlin Numeracy Scale, to assess an individual's level of risk literacy and statistical numeracy)
- [ ] Graph literacy: a cross-cultural comparison by M. Galesic and R. Garcia-Retamero: https://journals.sagepub.com/doi/10.1177/0272989X10373805

Done:
- mic-check
- color-blindness
- demographics (components to collect participant demographics): [vertical radio issue](https://github.com/revisit-studies/study/issues/506), ["other:__" issue](https://github.com/revisit-studies/study/issues/498)
- BeauVis: [randomize rows in Likert matrix](https://github.com/revisit-studies/study/issues/522)
- PreVis
- vlat

Unclear:
- [ ] Spatial ability test (I forget the name, but deals with paper folding): ?

### PR DESCRIPTION
**Pull request by JackWilb**
### Does this PR close any open issues?
Closes #478 

### Give a longer description of what this PR addresses and why it's needed
This adds libraries, a reusable configuration of components and sequences that can enable sharing of larger component blocks without copy/pasting between configurations.

The `LibraryConfig` is a new JSON schema that represents library objects. The schema looks similar to the `StudyConfig` (and reuses lots of the types), but it allows for multiple sequence definitions so that components may be used in 1 sequence and not another. This new schema is documented using typedoc.

### Provide pictures/videos of the behavior before and after these changes (optional)
New steps panel icon for library components.
Screenshot 2024-10-28 at 4 57 59 PM

### Are there any additional TODOs before this PR is ready to go?
TODOs:
- [x] Update relevant documentation
- [ ] Add a doc page to the docs repo
- [ ] Open issues for other libraries that we'd like to add

### CODE REVIEW

File: src/components/interface/StepsPanel.tsx [hunk 1]
```diff
@@ -123,6 +123,12 @@ function StepItem({
   const studyNavigateTo = () => (participantView ? navigate(`/${studyId}/${stepIndex}`) : navigate(`/${studyId}/reviewer-${step}`));
   const navigateTo = analysisNavigation ? analysisNavigateTo : studyNavigateTo;
 
+  // eslint-disable-next-line no-nested-ternary
+  const coOrComponents = step.includes('.co.')
+    ? 'co.'
+    : (step.includes('.components.') ? '.components.' : false);
+  const cleanedStep = step.includes('$') && coOrComponents && step.includes(coOrComponents) ? step.split(coOrComponents).at(-1) : step;
+
```
Comment (bbollen23) [inline src/components/interface/StepsPanel.tsx:131]: Are you basically allowing a user to use "$test.components.markdown-component" OR "$test.co.markdown-component"?? If so, is this necessary? Also, should the check have '.co.' on line 128? Don't think it actually ends up affecting the final cleanedStep, but at least for clarity.

**Review by bbollen23 — APPROVED** [review]
Just want to comment on this so I can see if I understand:

You have a config in your library. This is not initially namespaced -- it is just as if the user added their own config they used for a study there (to my knowledge, there's currently no restriction on the library elements, so any config thats valid should be able to be placed there).

When a user specifies a library, it firsts generates a "new config" which takes all the components and sequence objects and adds them as baseComponents with the specified namespace name. For example, if there is a baseComponent called "test-baseComponent", then a new base component in this new config is created with the name $test.components.test-baseComponent". That way, when you are referencing a component in your config, it has something to reference. Its essentially "merging" the library config inside your own config, but namespaces the elements.

This also allows for inheriting the libraryData's components as components or as baseComponents. 

So, couple questions:
1) Whats the use of a base component in the library config? is it just so that the library config itself can use them? Since it doesnt seem like you're adding specific libraryBaseComponents to the baseComponents of the final config, you're inheriting inside the library first? In other words, you're first generating the full library config so that there are no more "baseComponents" -- each final component of the resulting library config will essentially have all the stand-alone information for that component. (So if a library component inherits a library base component, when the library component is added to the final config, its as if that single library component has all the contents of the library base component coming with it)

2. Are there any checks in place for repeating library names in the library folder? Not sure how you'd handle this if you haven't already

Great work!! I like the namespace conventions you used

File: src/components/interface/StepsPanel.tsx (hunk 1 above)
Comment (JackWilb) [reply src/components/interface/StepsPanel.tsx]: Fixed in [6ac848e](https://github.com/revisit-studies/study/pull/488/commits/6ac848e580ad73fe31ed9e0439720d02bf4c62a8)

File: src/parser/parser.ts [hunk 2]
```diff
@@ -185,16 +191,28 @@ export function parseStudyConfig(fileData: string): ParsedStudyConfig {
...
-  let warnings: Required<ParsedStudyConfig>['warnings'] = [];
+  let errors: Required<ParsedConfig<StudyConfig>>['errors'] = studyValidate.errors || [];
+  let warnings: Required<ParsedConfig<StudyConfig>>['warnings'] = [];
 
   // We can only run our custom validator if the schema validation passes
   if (validatedData) {
-    const { errors: parserErrors, warnings: parserWarnings } = verifyStudyConfig(data!);
-    errors = [...(studyValidate.errors || []), ...parserErrors];
-    warnings = parserWarnings;
+    const importedLibraries = data!.importedLibraries || [];
```
Comment (ZachCutler04) [inline src/parser/parser.ts:199]: Is there some way for us to avoid having to force (!) the data object, some type check above?

**Review by ZachCutler04 — APPROVED** [review]
One small potential change, otherwise looks great

File: src/parser/parser.ts (hunk 2 above)
Comment (JackWilb) [reply src/parser/parser.ts]: Fixed in 60865a3

**Pull request merged**